### DISCUSSION TITLE
Unable to load custom plugin in strongswan

### DISCUSSION THREAD

**Discussion by deeptishm** [strongswan/strongswan#1032] [Q&A] [answered]
Hi,

I am trying to add a NEW plugin to strongswan ; used this as ref:
https://git.strongswan.org/?p=strongswan.git;a=shortlog;h=refs/heads/demo-plugin
The plugin libs were built successfully; however the plugin is not listed in ipsec statusall.

Can someone please guide or point out what I am missing here.

Attached are the detailed steps and logs:

Thanks
[Plugin-Issue.txt](https://github.com/strongswan/strongswan/files/8618415/Plugin-Issue.txt)
.

**Comment by tobiasbrunner** [comment]
Since the _kernel-netlink_ plugin is loaded even after you disabled it, I guess you installed strongSwan previously with different options. So you might want to remove all previous files (including the config files) before reinstalling a build that uses different config options.

**Reply by deeptishm** [reply]
I removed all previous files (including config files), but still unable to load the plugin in ipsec statusall
Attached are the details
[Plugin-Issue-resp.txt](https://github.com/strongswan/strongswan/files/8619712/Plugin-Issue-resp.txt)
.

**Reply by tobiasbrunner** [reply]
Increase the log level for `lib` to 3 and read the log when the daemon starts.

**Reply by deeptishm** [reply]
Increased log level of LIB to 3, do not see any reference to custom plugin.
Log Attached
[charon.log](https://github.com/strongswan/strongswan/files/8629185/charon.log)
.

**Reply by tobiasbrunner** [reply]
Does `libxilinx_plugin.c` actually contain a (non-static) `libxilinx_plugin_create()` function?

**Reply by tobiasbrunner** [reply]
I just pushed https://github.com/strongswan/strongswan/commit/af9c78d393882f8732f5ba558029d28e4882f0b0, which adds an explicit error message if the constructor is not found.

**Comment by deeptishm** [comment, marked as answer]
Since I copied the kernel_libipsec plugin to create custom plugin it had function of the name "kernel_libipsec_plugin_create()" , I changed it to libxilinx_plugin_create ```and now the plugin is getting loaded but unloading due to some error. I am working on fixing it.

stack@stack-Standard-PC-i440FX-PIIX-1996:~$ grep libxilinx /var/log/charon.log
May  5 13:20:21 00[LIB] plugin 'libxilinx': loaded successfully
May  5 13:20:21 00[LIB] loading feature CUSTOM:libxilinx in plugin 'libxilinx'
May  5 13:20:21 00[LIB] feature CUSTOM:libxilinx in plugin 'libxilinx' failed to load
May  5 13:20:21 00[LIB] unloading plugin 'libxilinx' without loaded features
stack@stack-Standard-PC-i440FX-PIIX-1996:~$

Thanks @tobiasbrunner a lot for your inputs.

**Reply by deeptishm** [reply]
Able to load custom plugin successfully.
`stack@stack-Standard-PC-i440FX-PIIX-1996:~$ grep libxilinx /var/log/charon.log
May  5 14:03:34 00[LIB] plugin 'libxilinx': loaded successfully
May  5 14:03:34 00[LIB]       loading feature CUSTOM:kernel-ipsec in plugin 'libxilinx'
May  5 14:03:34 00[LIB] loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem fips-prf gmp curve25519 xcbc cmac hmac drbg attr libxilinx kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
stack@stack-Standard-PC-i440FX-PIIX-1996:~$
`

**Reply by deeptishm** [reply]
Thanks for your time @tobiasbrunner
